Add new FO hook for additional content in cart product lines

ndabene opened the pull request:
The new hook `actionCartProductLineAdditionalContent` introduces enhanced flexibility for product display in the cart, offering significant advantages for both theme developers and module creators:

### Key Benefits

#### For Theme Developers
- Ability to customize cart product presentation without core modifications
- Enhanced control over product information layout in the cart
- Possibility to add theme-specific product details (e.g., custom badges, labels)
- Better integration with various design systems and frameworks
- Simplified responsive design implementation for cart items

#### For Module Developers
- Direct integration point for product-specific features
- Ability to display:
  - Real-time stock status
  - Delivery time estimates
  - Product customization options
  - Warranty information
  - Cross-selling suggestions
  - Product-specific promotions
  - Custom attributes and features

### Technical Implementation Advantages
- Clean separation of concerns
- No override requirements
- Standardized integration point
- Performance-optimized implementation
- Full access to product and cart context

### Use Cases Examples
1. Stock Management Modules
   - Display real-time inventory levels
   - Show warehouse location
   - indicate restock dates

2. Shipping Modules
   - Show delivery time estimates
   - Display shipping restrictions
   - Present handling requirements

3. Product Enhancement Modules
   - Display product certifications
   - Show eco-friendly badges
   - Present product ratings

This enhancement significantly improves PrestaShop's extensibility while maintaining clean code architecture and providing developers with more tools to create enhanced shopping experiences.
| Questions | Answers |
| --------- | ------- |
| Branch? | develop |
| Description? | Introduced a new hook, `actionCartProductLineAdditionalContent`, to allow for the inclusion of additional content for each product line in the cart. This enhances the cart's extensibility by enabling developers to customize the display of product information. |
| Type? | improvement |
| Category? | FO |
| BC breaks? | no |
| Deprecations? | no |
| How to test? | 1. Install the module [qahookactioncartproductlineadditionalcontent.zip](https://github.com/user-attachments/files/19531434/qahookactioncartproductlineadditionalcontent.zip)<br><br>2. In `themes/the_current/templates/checkout/_partials/cart-detailed-product-line.tpl` after:<br><pre>{block name='hook_cart_extra_product_actions'}<br>    {hook h='displayCartExtraProductActions' product=$product}<br>{/block}</pre><br>Add:<br><pre>{if isset($product.additional_info)}<br>    &lt;div class="product-line__additional-info"&gt;<br>        {foreach from=$product.additional_info key="key" item="info"}<br>            {foreach from=$info item="value"}<br>                &lt;div class="product-line__additional-info-item"&gt;<br>                    &lt;span class="label"&gt;{$key}:&lt;/span&gt;<br>                    &lt;span class="value"&gt;{$value}&lt;/span&gt;<br>                &lt;/div&gt;<br>            {/foreach}<br>        {/foreach}<br>    &lt;/div&gt;<br>{/if}</pre> |
| Fixed issue or discussion? | https://github.com/PrestaShop/PrestaShop/discussions/38368 |

ndabene committed:
Add hook for additional content in cart product lines

Introduced a new hook, `displayCartProductLineAdditionalContent`, to allow for the inclusion of additional content for each product line in the cart. This enhances the cart's extensibility by enabling developers to customize the display of product information.

ps-jarvis commented:
<!-- PR_WELCOME -->
Hello @ndabene!

This is your first pull request on [PrestaShop](https://github.com/PrestaShop/PrestaShop/) repository of the PrestaShop project.

Thank you, and welcome to this Open Source community!

ps-jarvis commented:
<!-- PR_TABLE_DESCRIPTION_ERROR -->
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `deprecations` should be `yes` or `no`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#deprecations))
* The `How to test` shouldn&#039;t be empty. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#how-to-test))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

PululuK commented on classes/Cart.php:
```diff
@@ -845,6 +845,16 @@ public function getProducts($refresh = false, $id_product = false, $id_country =
                 $this->_products[] = $product;
+
+                // Add hook for additional content on each cart product line
+                $additionalContent = Hook::exec('displayCartProductLineAdditionalContent', [
+                    'product' => $product,
+                    'cart' => $this
+                ], null, true);
+
+                if (!empty($additionalContent)) {
+                    $product['additional_content'] = $additionalContent;
+                }
```
Hi @ndabene , thank you for your contribution (ça fait plaisir )

However, I believe this hook is not placed optimally for the following reasons:

- The `Cart::getProducts` method is used in multiple contexts and is not intended to expose data directly to the front. Adding extra information here could introduce unnecessary processing, potentially impacting performance.

- Data intended for the front is handled by `CartPresenter::present,` where a hook (actionPresentCart) already exists, allowing you to override the data or add specific processing.

I believe the most appropriate place to add your hook is in `CartLazyArray::getProducts`, as it directly structures the cart data before being sent to the front. 

Also note that this is not a display type hook, but an action type hook

ndabene replied:
My idea is to be able add additionnal content everywhere you manipulate Cart::getProducts to add for example Present product ratings for reassurance or Show eco-friendly badges ( like energy label etc ). But I understand your feedback 👍

PululuK replied:
@ndabene you can use `CartLazyArray::getProducts` in the products loop

ndabene replied:
I'm not sure understand well :)

ndabene replied:
I got some new idea to improve this. I will update the PR asap

Hlavtox commented:
I like this, but I would go for a simple display hook in the template.

ndabene committed:
Remove additional content hook from cart product lines

Eliminated the previously introduced hook `displayCartProductLineAdditionalContent` from the cart product lines. This change simplifies the cart's product handling by removing the additional content functionality, streamlining the codebase.

ndabene committed:
Rename and repurpose hook for cart product lines

Updated the hook from `displayCartProductLineAdditionalContent` to `actionCartProductLineAdditionalContent`, enhancing its functionality to allow the addition of extra data and content to products in the shopping cart. This change improves the extensibility of the cart system.

ndabene committed:
Add hook execution for additional content in cart product lines

Enhanced the CartLazyArray class by executing the new hook 'actionCartProductLineAdditionalContent' for each product in the cart. This allows modules to add extra content to product lines, improving the extensibility and customization of the cart system.

ndabene committed:
Removed bad \ on Hook::exec

Hlavtox commented:
@ndabene You can do the same thing with `actionPresentCart` and `actionPresentProduct` no?

ndabene commented:
@Hlavtox ah yes my bad I just made a test on my dev module . So is a fail contribution for my 1st time. But if you have some needs on hook, you can ping me, I ll try find time to contribute around this

kpodemski commented:
@ndabene 

no worries, thanks for trying! if you are unsure about the hook you want to add, you can always open a discussion [here](https://github.com/PrestaShop/PrestaShop/discussions) and it is always a great idea to[ join us on Slack](https://www.prestashop-project.org/slack/), where we discuss a lot of things about PrestaShop

kpodemski commented:
I'm closing this PR because of the existing valid alternative in the core.
